packages/relayer/proof/encoded_signal_proof.go

```diff
@@ -47,16 +67,90 @@ func (p *Prover) EncodedSignalProof(
 	return encodedSignalProof, nil
 }
 
-// getEncodedStorageProof rlp and abi encodes a proof for LibBridgeSignal,
-// where `proof` is an rlp and abi encoded (bytes, bytes) consisting of the accountProof and storageProof.Proofs[0]
-// response from `eth_getProof`
+func (p *Prover) EncodedSignalProofWithHops(
+	ctx context.Context,
+	caller relayer.Caller,
+	signalServiceAddress common.Address,
+	hopParams []HopParams,
+	key string,
+	blockHash common.Hash,
+	blockNum uint64,
+) ([]byte, uint64, error) {
+	blockHeader, err := p.blockHeader(ctx, p.blocker, blockHash)
+	if err != nil {
+		return nil, 0, errors.Wrap(err, "p.blockHeader")
+	}
+
+	encodedStorageProof, signalRoot, err := p.encodedStorageProof(
+		ctx,
+		caller,
+		signalServiceAddress,
+		key,
+		blockHeader.Height.Int64(),
+	)
+
+	if err != nil {
+		return nil, 0, errors.Wrap(err, "p.encodedStorageProof")
+	}
+
+	slog.Info("successfully generated main storage proof proof")
+
+	hops := []encoding.Hop{}
+
+	for _, hop := range hopParams {
+		hopStorageSlotKey, err := hop.SignalService.GetSignalSlot(&bind.CallOpts{},
+			hop.ChainID,
+			hop.TaikoAddress,
+			signalRoot,
+		)
+		if err != nil {
+			return nil, 0, errors.Wrap(err, "hopSignalService.GetSignalSlot")
+		}
+
+		encodedHopStorageProof, nextSignalRoot, err := p.encodedStorageProof(
+			ctx,
+			hop.Caller,
+			hop.SignalServiceAddress,
+			common.Bytes2Hex(hopStorageSlotKey[:]),
+			int64(blockNum),
```

Comment:
upon reviewing my own code, this only works for one hop, we need to add `blockNum` as a `hopParam`.

Reply:
resolved now